salt/modules/system.py

```diff
@@ -608,3 +608,54 @@ def get_computer_name():
         salt '*' network.get_hostname
     '''
     return __salt__['network.get_hostname']()
+
+
+NILRT_REBOOT_WITNESS_PATH = '/var/volatile/tmp/salt/reboot_witnessed'
+
+
+def set_reboot_required_witnessed():
+    '''
+    This function is used to remember that an event indicating that a reboot is
+    required was witnessed. This function writes to a temporary filesystem so
+    the event gets cleared upon reboot.
+
+    Returns:
+        bool: ``True`` if successful, otherwise ``False``
+
+    .. code-block:: bash
+
+        salt '*' system.set_reboot_required_witnessed
+    '''
+    if __grains__.get('os_family') == 'NILinuxRT':
+        dir_path = os.path.dirname(NILRT_REBOOT_WITNESS_PATH)
+        if not os.path.exists(dir_path):
+            try:
+                os.makedirs(dir_path)
+            except OSError as ex:
+                raise SaltInvocationError('Error creating {0} (-{1}): {2}'
+                                          .format(dir_path, ex.errno, ex.strerror))
+
+        __salt__['cmd.run']('touch {0}'.format(NILRT_REBOOT_WITNESS_PATH))
+        return True
+
+    return False
+
+
+def get_reboot_required_witnessed():
+    '''
+    Determine if at any time during the current boot session the salt minion
+    witnessed an event indicating that a reboot is required.
+
+    Returns:
+        bool: ``True`` if the a reboot request was witnessed, ``False`` otherwise
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' system.get_reboot_required_witnessed
+    '''
+    if __grains__.get('os_family') == 'NILinuxRT':
+        if os.path.exists(NILRT_REBOOT_WITNESS_PATH):
+            return True
+    return False
```

Comment:
Actually, I like the idea of such decorators even if we don't use them here, so I am looking into writing them now.

Reply:
@terminalmage :+1: for the decorator. I would still map it separately into its own namespace. Because it is "special" stuff. Then in your state you can better segregate these calls, like:

```
salt * somemodule.special.do_something
```
Then you will also construct your states/commands differently, avoiding hitting systems that aren't in this space.

Reply:
@skizunov are we on the same page? Sounds like you're talking about an interface. And it is yet to be implemented. What I am talking is a principal of an actual implementation of _any_ module that on some OSes even do not have to exist. That said, if other systems needs to implement this, then exception raise "not yet implemented" or "cannot be at all supported" is needed here.

But right now it is returning one of an identical return data of the actual implementation. And so this is a bug and confusion.

Reply:
@isbm : This interface has already been implemented for Windows systems (in `win_system.py` which also implements the `system` virtual interface). But I understand your point about the behavior on other Linux distributions that may not necessarily implement the interface. So perhaps there are a few options:
1) Use the solution that @terminalmage suggests (requires infrastructure not yet present)
2) Just implement for the other Linux distributions (should be very easy, just need to confirm the path to volatile storage is valid for each Linux distribution)
3) On other Linux distributions that don't (yet) support this, throw an exception instead of returning a value.

Reply:
@skizunov I'd prefer @terminalmage 's suggestion as it is case-agnostic.

Reply:
Implementing function namespaces is not going to happen right now. It may in a future release, but it's extremely unlikely that it will happen in time for Fluorine. So, we should adjust expectations here accordingly.

Late last week I added the needed support to the `depends` decorator to enable the running of a shell command (https://github.com/saltstack/salt/pull/47040). In the process of doing so, I discovered that this decorator _already_ supports interpreting an expression that evaluates to a boolean (with `True` meaning that the function should be loaded and `False` meaning that it should not).

There are two approaches we could use:

**OPTION 1 (depends on https://github.com/saltstack/salt/pull/47040, would have to wait for that PR to be merged and then rebase to gain access to this functonality)**

```python
from salt.utils.decorators import depends

@depends('some shell command', retcode=0)
def set_reboot_required_witnessed():
    '''
```

**OPTION 2**

```python
from salt.utils.decorators import depends

@depends(some python expression)
def set_reboot_required_witnessed():
    '''
```

Option 2 could be done right now without waiting for my decorator modifications to be merged, but note that whatever is passed to the depends decorator must evaluate as boolean. So it has to be an expression or function that returns a bool.

The question is, I guess, what would be the best way to confirm that the minion is NILinuxRT? Keep in mind that we won't have access to the grains.

Reply:
I've rebased on top of master and have the code from #47040 

Can I run any shell commands in the depends decorator? I can check if we're on NILRT using 'lsb-release -i | grep -q nilrt ' if yes.